File: agents/java/afl-agent/src/main/java/afl/agent/MongoOps.java
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package afl.agent;

import afl.agent.model.StepAttribute;
import afl.agent.model.TaskDocument;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.FindOneAndUpdateOptions;
import com.mongodb.client.model.ReturnDocument;
import com.mongodb.client.model.Updates;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static com.mongodb.client.model.Filters.*;

/**
 * MongoDB operations for the AFL agent protocol.
 */
public class MongoOps {

    private final MongoDatabase db;

    public MongoOps(MongoDatabase db) {
        this.db = db;
    }

    /**
     * Atomically claims a pending task for processing.
     * Returns empty if no task is available.
     */
    public Optional<TaskDocument> claimTask(List<String> taskNames, String taskList) {
        MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_TASKS);

        Bson filter = and(
                eq("state", Protocol.TASK_STATE_PENDING),
                in("name", taskNames),
                eq("task_list_name", taskList)
        );

        Bson update = Updates.combine(
                Updates.set("state", Protocol.TASK_STATE_RUNNING),
                Updates.set("updated", nowMillis())
        );

        FindOneAndUpdateOptions options = new FindOneAndUpdateOptions()
                .returnDocument(ReturnDocument.AFTER);

        Document doc = collection.findOneAndUpdate(filter, update, options);

        if (doc == null) {
            return Optional.empty();
        }

        return Optional.of(TaskDocument.fromDocument(doc));
    }

    /**
     * Reads the params attribute from a step.
     */
    public Map<String, Object> readStepParams(String stepId) {
        MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_STEPS);

        Document step = collection.find(eq("uuid", stepId)).first();
        if (step == null) {
            throw new IllegalArgumentException("Step not found: " + stepId);
        }

        Map<String, Object> result = new HashMap<>();

        Document attributes = step.get("attributes", Document.class);
        if (attributes != null) {
            Document params = attributes.get("params", Document.class);
            if (params != null) {
                for (String name : params.keySet()) {
                    Document attr = params.get(name, Document.class);
                    if (attr != null) {
                        result.put(name, attr.get("value"));
                    }
                }
            }
        }

        return result;
    }

    /**
     * Writes return attributes to a step.
     */
    public void writeStepReturns(String stepId, Map<String, Object> returns) {
        MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_STEPS);

        Document setFields = new Document();
        for (Map.Entry<String, Object> entry : returns.entrySet()) {
            String name = entry.getKey();
            Object value = entry.getValue();

            Document attr = new Document()
                    .append("name", name)
                    .append("value", value)
                    .append("type_hint", StepAttribute.inferTypeHint(value));

            setFields.append("attributes.returns." + name, attr);
        }

        collection.updateOne(
                and(
                        eq("uuid", stepId),
                        eq("state", Protocol.STEP_STATE_EVENT_TRANSMIT)
                ),
                new Document("$set", setFields)
        );
    }

    /**
     * Marks a task as completed.
     */
    public void markTaskCompleted(TaskDocument task) {
        MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_TASKS);

        collection.updateOne(
                eq("uuid", task.uuid()),
                Updates.combine(
                        Updates.set("state", Protocol.TASK_STATE_COMPLETED),
                        Updates.set("updated", nowMillis())
                )
        );
    }

    /**
     * Marks a task as failed with an error message.
     */
    public void markTaskFailed(TaskDocument task, String errorMsg) {
        MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_TASKS);

        collection.updateOne(
                eq("uuid", task.uuid()),
                Updates.combine(
                        Updates.set("state", Protocol.TASK_STATE_FAILED),
                        Updates.set("updated", nowMillis()),
                        Updates.set("error", new Document("message", errorMsg))
                )
        );
    }

    /**
     * Inserts a step log entry for dashboard observability.
     * Best-effort: errors are caught and logged.
     */
    public void insertStepLog(String stepId, String workflowId, String runnerId,
                              String facetName, String source, String level, String message) {
        try {
            MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_STEP_LOGS);

            long now = nowMillis();
            Document doc = new Document()
                    .append("uuid", UUID.randomUUID().toString())
                    .append("step_id", stepId)
                    .append("workflow_id", workflowId)
                    .append("runner_id", runnerId)
                    .append("facet_name", facetName)
                    .append("source", source)
                    .append("level", level)
                    .append("message", message)
                    .append("details", new Document())
                    .append("time", now);

            collection.insertOne(doc);
        } catch (Exception e) {
            java.util.logging.Logger.getLogger(MongoOps.class.getName())
                    .fine("Could not save step log for step " + stepId + ": " + e.getMessage());
        }
    }

    /**
     * Inserts an afl:resume task for the Python RunnerService.
     */
    public void insertResumeTask(String stepId, String workflowId, String taskList) {
        MongoCollection<Document> collection = db.getCollection(Protocol.COLLECTION_TASKS);

        long now = nowMillis();
        Document task = new Document()
                .append("uuid", UUID.randomUUID().toString())
                .append("name", Protocol.RESUME_TASK_NAME)
                .append("runner_id", "")
                .append("workflow_id", workflowId)
                .append("flow_id", "")
                .append("step_id", stepId)
                .append("state", Protocol.TASK_STATE_PENDING)
                .append("created", now)
                .append("updated", now)
                .append("task_list_name", taskList)
                .append("data_type", "resume")
                .append("data", new Document()
                        .append("step_id", stepId)
                        .append("workflow_id", workflowId));

        collection.insertOne(task);
    }

    /**
     * Returns current time in milliseconds since Unix epoch.
     */
    public static long nowMillis() {
        return System.currentTimeMillis();
    }
}

File: agents/scala/fw-agent/src/main/scala/afl/agent/MongoOps.scala
package fw.agent

import fw.agent.model.{StepAttributes, TaskDocument}
import fw.agent.model.TaskDocument.*
import org.mongodb.scala.*
import org.mongodb.scala.bson.{BsonNull, Document}
import org.mongodb.scala.model.{Filters, FindOneAndUpdateOptions, ReturnDocument, Updates}
import org.slf4j.LoggerFactory

import java.util.UUID
import scala.concurrent.Await
import scala.concurrent.duration.*
import scala.jdk.CollectionConverters.*

/** MongoDB operations for the AFL agent poller.
  *
  * All operations are synchronous (blocking) to match the polling model.
  */
class MongoOps(db: MongoDatabase, timeout: Duration = 10.seconds):

  private val logger = LoggerFactory.getLogger(getClass)
  private val tasks = db.getCollection(Protocol.Collections.Tasks)
  private val steps = db.getCollection(Protocol.Collections.Steps)

  /** Atomically claim a pending task (pending -> running).
    *
    * @param taskNames  Event facet names to match
    * @param taskList   Task list name for routing
    * @return           The claimed task, or None if no tasks available
    */
  def claimTask(
      taskNames: Seq[String],
      taskList: String
  ): Option[TaskDocument] =
    val filter = Filters.and(
      Filters.eq("state", Protocol.TaskState.Pending),
      Filters.in("name", taskNames*),
      Filters.eq("task_list_name", taskList)
    )
    val update = Updates.combine(
      Updates.set("state", Protocol.TaskState.Running),
      Updates.set("updated", System.currentTimeMillis())
    )
    val options = FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER)

    val result = Await.result(
      tasks.findOneAndUpdate(filter, update, options).toFuture(),
      timeout
    )
    Option(result).map(doc => TaskDocument.fromBson(Document(doc.toBsonDocument)))

  /** Read step params from the steps collection.
    *
    * @param stepId  The step UUID
    * @return        Map of param name to value
    */
  def readStepParams(stepId: String): Map[String, Any] =
    val filter = Filters.eq("uuid", stepId)
    val result =
      Await.result(steps.find(filter).first().toFuture(), timeout)
    Option(result) match
      case Some(doc) =>
        val stepDoc = Document(doc.toBsonDocument)
        StepAttributes
          .extractParams(stepDoc)
          .map((k, v) => k -> v.value)
      case None =>
        logger.warn(s"Step not found: $stepId")
        Map.empty

  /** Write return attributes to a step at EVENT_TRANSMIT state.
    *
    * @param stepId   The step UUID
    * @param returns  Map of return name to (value, typeHint)
    */
  def writeStepReturns(
      stepId: String,
      returns: Map[String, (Any, String)]
  ): Unit =
    val filter = Filters.and(
      Filters.eq("uuid", stepId),
      Filters.eq("state", Protocol.StepState.EventTransmit)
    )
    val updates = returns.map { case (name, (value, typeHint)) =>
      Updates.set(
        s"attributes.returns.$name",
        Document(
          "name" -> name,
          "value" -> scalaToMongo(value),
          "type_hint" -> typeHint
        )
      )
    }.toSeq

    if updates.nonEmpty then
      Await.result(
        steps.updateOne(filter, Updates.combine(updates*)).toFuture(),
        timeout
      )

  /** Merge partial return attributes into a step.
    * Unlike writeStepReturns, this does NOT require the step to be in EVENT_TRANSMIT state,
    * allowing handlers to stream partial results during execution.
    *
    * @param stepId   The step UUID
    * @param partial  Map of return name to (value, typeHint)
    */
  def updateStepReturns(
      stepId: String,
      partial: Map[String, (Any, String)]
  ): Unit =
    val filter = Filters.eq("uuid", stepId)
    val updates = partial.map { case (name, (value, typeHint)) =>
      Updates.set(
        s"attributes.returns.$name",
        Document(
          "name" -> name,
          "value" -> scalaToMongo(value),
          "type_hint" -> typeHint
        )
      )
    }.toSeq

    if updates.nonEmpty then
      Await.result(
        steps.updateOne(filter, Updates.combine(updates*)).toFuture(),
        timeout
      )

  /** Mark an event task as completed. */
  def markTaskCompleted(task: TaskDocument): Unit =
    val completed = task.copy(
      state = Protocol.TaskState.Completed,
      updated = System.currentTimeMillis()
    )
    val filter = Filters.eq("uuid", task.uuid)
    Await.result(
      tasks.replaceOne(filter, completed.toBson).toFuture(),
      timeout
    )

  /** Mark an event task as failed with an error message. */
  def markTaskFailed(task: TaskDocument, errorMessage: String): Unit =
    val failed = task.copy(
      state = Protocol.TaskState.Failed,
      updated = System.currentTimeMillis(),
      error = Some(Document("message" -> errorMessage))
    )
    val filter = Filters.eq("uuid", task.uuid)
    Await.result(
      tasks.replaceOne(filter, failed.toBson).toFuture(),
      timeout
    )

  /** Insert an afl:resume task so the Python RunnerService resumes the workflow.
    *
    * @param stepId      The step whose returns have been written
    * @param workflowId  The workflow to resume
    * @param taskList    Task list name for routing
    */
  def insertResumeTask(
      stepId: String,
      workflowId: String,
      taskList: String,
      facetName: String = ""
  ): Unit =
    val nowMs = System.currentTimeMillis()
    val resumeName = if facetName.nonEmpty then s"${Protocol.ResumeTaskName}:$facetName" else Protocol.ResumeTaskName
    val resumeTask = TaskDocument(
      uuid = UUID.randomUUID().toString,
      name = resumeName,
      runnerId = "",
      workflowId = workflowId,
      flowId = "",
      stepId = stepId,
      state = Protocol.TaskState.Pending,
      created = nowMs,
      updated = nowMs,
      error = None,
      taskListName = taskList,
      dataType = "resume",
      data = Some(
        Document(
          "step_id" -> stepId,
          "workflow_id" -> workflowId
        )
      )
    )
    Await.result(tasks.insertOne(resumeTask.toBson).toFuture(), timeout)

  /** Insert a step log entry for dashboard observability.
    *
    * Best-effort: errors are caught and logged at debug level.
    */
  def insertStepLog(
      stepId: String,
      workflowId: String,
      runnerId: String,
      facetName: String,
      source: String,
      level: String,
      message: String
  ): Unit =
    try
      val nowMs = System.currentTimeMillis()
      val doc = Document(
        "uuid" -> UUID.randomUUID().toString,
        "step_id" -> stepId,
        "workflow_id" -> workflowId,
        "runner_id" -> runnerId,
        "facet_name" -> facetName,
        "source" -> source,
        "level" -> level,
        "message" -> message,
        "details" -> Document(),
        "time" -> nowMs
      )
      Await.result(
        db.getCollection(Protocol.Collections.StepLogs).insertOne(doc).toFuture(),
        timeout
      )
    catch
      case e: Exception =>
        logger.debug(s"Could not save step log for step $stepId: ${e.getMessage}")

  /** Convert a Scala value to a MongoDB-compatible BsonValue. */
  private def scalaToMongo(value: Any): org.mongodb.scala.bson.BsonValue =
    import org.mongodb.scala.bson.*
    value match
      case null          => BsonNull()
      case s: String     => BsonString(s)
      case i: Int        => BsonInt32(i)
      case l: Long       => BsonInt64(l)
      case d: Double     => BsonDouble(d)
      case b: Boolean    => BsonBoolean(b)
      case seq: Seq[?]   => BsonArray.fromIterable(seq.map(scalaToMongo))
      case m: Map[?, ?] =>
        val bsonDoc = new org.bson.BsonDocument()
        m.foreach { case (k, v) =>
          bsonDoc.put(k.toString, scalaToMongo(v))
        }
        bsonDoc
      case other => BsonString(other.toString)
